components/todoist/sources/completed-task/completed-task.js

```diff
@@ -0,0 +1,46 @@
+const todoist = require("../../todoist.app.js");
+
+module.exports = {
+  key: "todoist-completed-task",
+  name: "Completed Task",
+  description: "Emit an event for each completed task",
+  version: "0.0.1",
+  props: {
+    todoist,
+    selectProjects: { propDefinition: [todoist, "selectProjects"] },
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 5,
+      },
+    }, 
+    db: "$.service.db",
+  }, 
+  dedupe: "unique",
+  async run(event) {
+    const sync_token = this.db.get("sync_token") || '*'
+    const resourceTypes = ['items']
+
+    const result = await this.todoist.sync({
+      resource_types: JSON.stringify(resourceTypes),
+      sync_token,
+    })
+  
+    for (const property in result) {
+      if(Array.isArray(result[property])) {
+        for (const element of result[property]) {
+          let matchingProject = await this.todoist.isProjectInList(element.project_id, this.selectProjects)
+          if(element.checked === 1 && matchingProject) {
+            let dedupeId = `${element.id}-${(new Date(element.date_completed)).getTime()}`
+            this.$emit(element, {
+              summary: element.content,
+              id: dedupeId, 
+            })
```

Comment:
We should extract the metadata elements filtering and the "generation" into separate methods (e.g. `isElementRelevant`, `generateMeta`, etc.) so that we can further abstract this piece of code.

For example:
```js
// sources/common.js
methods: {
  generateMeta(element) {
    const {
      id: elementId,
      content: summary,
      date_completed: dateCompleted,
    } = element;
    const ts = new Date(dateCompleted).getTime();
    const id = `${elementId}-${ts}`;
    return {
      id,
      summary,
      ts,
    };
  }
},
async run() {
  const syncResult = await this.getSyncResult();
  Object.values(syncResult)
    .filter(Array.isArray)
    .flat()
    .filter(this.isElementRelevant)
    .filter(element => this.todoist.isProjectInList(element.project_id, this.selectProjects))
    .forEach(element => {
      const meta = this.generateMeta(element);
      this.$emit(element, meta);
    });
}

// sources/completed-task/completed-task.js
const common = require("../../common");
module.exports = {
  ...common,
  // Key, description, etc.
  methods: {
    isElementRelevant(element) {
      return element.checked === 1;
    },
  },
};
```

This way, we can limit each event source to only contain logic that is particular to its use case, and have the rest of the boilerplate in common modules.
I suggest having:
- `sources/common-task.js`
- `sources/common-project.js`
- `sources/common.js` (parent of the above)